c++/src/kj/common.h

```diff
@@ -1295,19 +1298,65 @@ class NullableValue {
   friend NullableValue<U>&& readMaybe(Maybe<U>&& maybe);
 };
 
+template <typename T, typename = EnableIf<!isLvalueReference<T>()>>
+class ExplicitPtr {
+  // This class wraps a raw pointer to T. It supports one-way implicit conversion: raw pointers can
+  // be implicitly converted to an ExplicitPtr, but an ExplicitPtr cannot be implicitly converted
+  // back to a raw pointer. The only operations ExplicitPtr supports are:
+  //
+  //   - Dereferencing with operator*.
+  //   - Indirection with operator->.
+  //   - Contextual conversion to bool with operator bool.
+  //
+  // If you require a raw pointer from an ExplicitPtr, you must use explicit syntax such as `&*p` or
+  // `p.operator->()`, or `static_cast<T*>(p)`.
+  //
+  // The purpose of this class is to provide greater compile-time safety for the variable declared
+  // by the `KJ_IF_MAYBE(p, maybe)` macro -- it is common for authors to forget that `p` is a
+  // pointer, and this confusion has been the frequent cause of subtle bugs with sometimes-severe
+  // consequences over the years. By making `p` an ExplicitPtr instead of a raw pointer,
+  // accidentally passing `p` to a function which accepts both a raw pointer and the pointed-to type
+  // (e.g., `kj::str()`, `kj::hashCode()`, and many more) becomes a compile-time error.
+  //
+  // This class is SFINAE-friendly, so that it may be used in function signatures wherever T* may be
+  // used.
+
+public:
+  ExplicitPtr(T* ptr): ref(ptr) {}
+  // A raw pointer can be implicitly converted to ExplicitPtr.
+
+  T& operator*() { return *ref; }
+  T* operator->() { return ref; }
+  const T& operator*() const { return *ref; }
+  const T* operator->() const { return ref; }
+  // ExplicitPtr supports pointer-like dereference/indirection operations.
+
+  explicit operator T*() { return ref; }
+  explicit operator const T*() const { return ref; }
+  // An ExplicitPtr may be explicitly converted back to its raw pointer.
+
+  explicit operator bool() const { return ref != nullptr; }
+  // An ExplicitPtr may be contextually converted to bool for use in `if` conditions.
```

Comment:
I guess this is needed only for the KJ_IF_MAYBE macro itself to function correctly?

If we're switching to C++17 anyway, can we instead rely on init-statement syntax, like:

```
if (ExplicitPtr<T> foo = blah(); foo.isNotNull()) {
```

?

Reply:
Oh geez I finally figured out how to make the identifier a reference, if we can assume C++17:

```
#define KJ_IF_MAYBE2(name, exp) \
  if (bool b = true; auto _##name = ::kj::_::readMaybe(exp)) \
    for (auto& name = *_##name; b; b = false)
```

If the compiler optimizes this reasonably it might be worth it?

Reply:
Holy shit https://godbolt.org/z/6x9zvsxqT

Reply:
> I guess this is needed only for the KJ_IF_MAYBE macro itself to function correctly?

Yes.

As for the reference hack, that's awesome! :) It looks like that even works correctly with `KJ_IF_MAYBE() { ... } else { ... }` constructions.

Personally, I'd be willing to replace every `KJ_IF_MAYBE` with the reference-declaring version (after upgrading to C++17/20, of course), since it's obviously better.

Reply:
Re: naming, how about `KJ_IF_NONNULL(var, maybe) { ... }`?

Reply:
I guess I could accept that though I also think that appending `2` to something, while aesthetically displeasing, has the nice effect of communicating that this is a new version of the thing, so people don't get confused about which one they should be using.